### ISSUE TITLE
should both: attention_mask and global_attention_mask be used for classification?

### ISSUE THREAD

**Issue by vr25** [allenai/longformer#75]
Hi,

Again a conceptual question on text classification.

Since global attention is used on ```<s>``` only, I am slightly confused if I should just pass global_attention_mask to the model or both: attention_mask and global_attention_mask. I follow that attention_mask is mainly used to mask the ```<pad>``` tokens but does it mean ``` n ```  <sup>2</sup> complexity for local attention?

Thanks!

**Comment by ibeltagy** [comment]
As the docstring [here](https://github.com/huggingface/transformers/blob/master/src/transformers/modeling_longformer.py#L457-L473) says: 
`attention_mask`: some attention or no attention
`global_attention_mask`: local attention or global attention

Check [here](https://github.com/huggingface/transformers/blob/master/src/transformers/modeling_longformer.py#L642-L652) for how we merge both masks into the {0, 1, 2} mask.

**Comment by vr25** [comment]
Why is [this](https://github.com/huggingface/transformers/blob/master/src/transformers/modeling_longformer.py#L861) 
not 

if global_attention_mask is **not** None: ?

**Comment by ibeltagy** [comment]
To answer your emailed questions: 

- `LongformerModel` and `LongformerForSequenceClassification`: both don't do n^2 attention
- `LongformerModel` and `LongformerForSequenceClassification`: the pretrained model supports seqlen up to 4096, but if you build your own, you can fit 64K tokens in a 48GB gpu with fp16 O2 and gradient checkpointing
- `LongformerModel` and `LongformerForSequenceClassification`: if you set `global_attention_mask` to True for all positions, this will be equivalent to n^2, but this is not how the model is supposed to be used.
- `LongformerModel`: default is local attention everywhere. You need to configure `global_attention_mask` yourself to add global attention to a few select tokens
- `LongformerForSequenceClassification` automatically sets `global_attention_mask` such that global attention is on the CLS token. You can override that behavior by providing your own `global_attention_mask`
- You can use the implantation in this repo or the one in the HF code under `https://github.com/huggingface/transformers/blob/master/src/transformers/modeling_longformer.py`

I am not sure if I answered all the questions, please let me know if I didn't.

**Comment by vr25** [comment]
Yes, thank you so much for answering all the questions I had so far.

**Comment by Yangxiaojun1230** [comment]
> To answer your emailed questions:
> 
> * `LongformerModel` and `LongformerForSequenceClassification`: both don't do n^2 attention
> * `LongformerModel` and `LongformerForSequenceClassification`: the pretrained model supports seqlen up to 4096, but if you build your own, you can fit 64K tokens in a 48GB gpu with fp16 O2 and gradient checkpointing
> * `LongformerModel` and `LongformerForSequenceClassification`: if you set `global_attention_mask` to True for all positions, this will be equivalent to n^2, but this is not how the model is supposed to be used.
> * `LongformerModel`: default is local attention everywhere. You need to configure `global_attention_mask` yourself to add global attention to a few select tokens
> * `LongformerForSequenceClassification` automatically sets `global_attention_mask` such that global attention is on the CLS token. You can override that behavior by providing your own `global_attention_mask`
> * You can use the implantation in this repo or the one in the HF code under `https://github.com/huggingface/transformers/blob/master/src/transformers/modeling_longformer.py`
> 
> I am not sure if I answered all the questions, please let me know if I didn't.

Hi guy，
     I meet an issue when I input only embed_ids and attention_mask . Do you know why？
When I train on a classification model by Longformer
def forward(self,input):
embding=input['enc']
att_mask=input['mask']
att_mask[:,[100,300,500,800,1200,]]=2
labels=input['targets']
print('jeff:',embding.device,att_mask.device,self.l1.device,embding.shape,att_mask.shape,self.maxlen)

logit=self.l1(inputs_embeds=embding,attention_mask=att_mask)#[:2]
    return [logit,labels]
Meet error:
pytorch/aten/src/THC/THCTensorIndex.cu:361: void indexSelectLargeIndex(TensorInfo<T, IndexType>, TensorInfo<T, IndexType>, TensorInfo<long, IndexType>, int, int, IndexType, IndexType, long) [with T = float, IndexType = unsigned int, DstDim = 2, SrcDim = 2, IdxDim = -2, IndexIsMajor = true]: block: [205,0,0], thread: [95,0,0] Assertion srcIndex < srcSelectDimSize failed

max_num_extra_indices_per_batch = num_extra_indices_per_batch.max()
RuntimeError: cuda runtime error (710) : device-side assert triggered at /pytorch/aten/src/THC/THCReduceAll.cuh:327

I checked the length of the attention_mask is same as the config.max_len which is like [bs,max_len]
Do anyone meet the same issue?

**Comment by ibeltagy** [comment]
Can you run your code again with `CUDA_LAUNCH_BLOCKING=1` to get a more accurate error message and post the full stack trace? Also, it is helpful if you can post a short snippet that reproduces your error.

**Comment by ibeltagy** [comment]
This question is a duplicate of https://github.com/allenai/longformer/issues/75. Closing in favor of the other one.

**Comment by mihaidobri** [comment]
@ibeltagy which is the duplicate of this question? Your last comment indicates probably by accident #75 which is actually the question topic above. Thank you!

**Comment by ibeltagy** [comment]
@mihaidobri, you are right, sorry, reopened it.

**Comment by pul95** [comment]
I am continuously getting the runtime error: CUDA error: device-side assert triggered, I am new to transformer library.  

I am creating the longformer classifier in the below format:

model = LongformerForSequenceClassification.from_pretrained('allenai/longformer-base-4096', gradient_checkpointing = True, num_labels = 5)

and tokenizer as :

tokenizer = LongformerTokenizer.from_pretrained('allenai/longformer-base-4096')

encoded_data_train = tokenizer.batch_encode_plus(
    df[df.data_type=='train'].content.values,
    add_special_tokens = True,
    return_attention_mask=True,
    padding = True,
    max_length = 3800,
    return_tensors='pt'
)

Can anyone please help? I am using latest version of transformer library and using google colab GPU for building my classifier.

**Comment by thesby** [comment]
If I use default attention config, the model just runs with local attention, right? So if my model `max_length=4096` and attention window is 512, each row of attention_mask will contains 512 ones near diagonal, right?

**Comment by pratikchhapolika** [comment]
> here

the page doesn't exist
